msal/src/Microsoft.Identity.Client/TokenCache.cs

```diff
@@ -134,6 +134,13 @@ internal void OnBeforeWrite(TokenCacheNotificationArgs args)
 
             IdToken idToken = IdToken.Parse(response.IdToken);
 
+            //injecting tenantID into the IDToken if it is null
+            if (requestParams.Authority.AuthorityType == Core.Instance.AuthorityType.B2C)
+            {
+                if (string.IsNullOrEmpty(idToken.TenantId) && !string.IsNullOrEmpty(tenantId))
+                    idToken.TenantId = tenantId;
+            }
+
             var msalAccessTokenCacheItem =
                 new MsalAccessTokenCacheItem(preferredEnvironmentHost, requestParams.ClientId, response, tenantId)
```

Comment:
it's probably me reading the code wrong, but why do we pass the "tenantId" from request params authority to the MsalAccessTokenCacheItem?

Reply:
The MsalAccessTokenCacheItem represents a data entry within the token cache. This item has a function called GetKey() which will generate a key for the token cache dictionary to use with the tenantID and other data within the MsalAccessTokenCacheItem. 
see: [MsalAccessTokenCacheItem Line 114](https://github.com/AzureAD/azure-activedirectory-library-for-dotnet/blob/a25be22e9d667050b96eb4517f1401c2dd97a119/core/src/Cache/MsalAccessTokenCacheItem.cs#L114), [TokenCacheAccessor Line 57](https://github.com/AzureAD/azure-activedirectory-library-for-dotnet/blob/a25be22e9d667050b96eb4517f1401c2dd97a119/core/src/Features/Cache/TokenCacheAccessor.cs#L57)

Reply:
Yes, that makes sense. However I was referring to where the "tenantId" is coming from:

`var tenantId = Authority.CreateAuthority(requestParams.TenantUpdatedCanonicalAuthority, false).GetTenantId();`

To me it seems that it's coming from developer requested authority and not idToken.TenantId.

Reply:
The requestParams.TenantUpdatedCanonicalAuthority is set by the id token. what is happening here is the requestParams is updated with the new authority returned from the idToken in case the user uses the common authority for authentication. 

Thus, https://login.microsoftonline.com/common/ will be replaced with https://login.microsoftonline.com/SomeTenantID/... in the requestParams.TenantUpdatedCanonicalAuthority property

[See RequestBase.cs Line 228](https://github.com/AzureAD/azure-activedirectory-library-for-dotnet/blob/e38e4073e9cc6b3c986feb0572286ab66dbb4202/msal/src/Microsoft.Identity.Client/Internal/Requests/RequestBase.cs#L228)